File: C++/Binary-Tree/Balanced Binary Tree.cpp
/*Given a binary tree, determine if it is height-balanced.
For this problem, a height-balanced binary tree is defined as:
A binary tree in which the depth of the two subtrees of every node never differ by more than 1.
Example 1:

Given the following tree [3,9,20,null,null,15,7]:

    3
   / \
  9  20
    /  \
   15   7
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        return (height_difference(root)!=-1);
        
    }
    private:
        int height_difference(TreeNode* root)
        {
            if(root == NULL)
                return 0;
            int left_tree_height = height_difference(root->left);
            if(left_tree_height == -1)
                return -1;
            
            int right_tree_height = height_difference(root->right);
            if(right_tree_height == -1)
                return -1;
            
            if(abs(left_tree_height - right_tree_height)>1)
                return -1;
            
            return(max(left_tree_height,right_tree_height)+1);
            
        }
};

File: Meta/Practice/Python/Balanced Binary Tree.py
'''
Given a binary tree. Determine if its height balanced or not.
Balanced 
Given the definition of a balanced tree
we know that a tree T is not balanced if and only if there is some node
p∈T  such that ∣height(p.left)−height(p.right)∣>1
The tree below has each node is labeled by its height,
as well as the unbalanced subtree highlighted.
'''
'''
Input: root = [3,9,20,null,null,15,7]
Output: true
'''

'''
Complexity:-
Tc:- O(n) 
where n is the number of nodes
Sc:- O(n)
for the recursion stack
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def heightDifference(root):
            if not root:
                return 0
            leftTreeHeight = heightDifference(root.left)
            if leftTreeHeight == -1:
                return -1
            rightTreeHeight = heightDifference(root.right)
            if rightTreeHeight == -1:
                return -1
            if abs(leftTreeHeight - rightTreeHeight) > 1:
                return -1
            return (max(leftTreeHeight,rightTreeHeight)+1)

        return heightDifference(root) != -1 
        
